plugins/speculation-rules/load.php

```diff
@@ -20,15 +20,67 @@
 	exit;
 }
 
-// Define the constant.
-if ( defined( 'SPECULATION_RULES_VERSION' ) ) {
-	return;
-}
+(
+	/**
+	 * Register this copy of the plugin among other potential copies.
+	 *
+	 * @param string  $global_var_name Global variable name for storing the plugin pending loading.
+	 * @param string  $version         Version.
+	 * @param Closure $load            Callback that loads the plugin.
+	 */
+	static function ( string $global_var_name, string $version, Closure $load ) {
+		if ( ! isset( $GLOBALS[ $global_var_name ] ) ) {
+			$bootstrap = static function () use ( $global_var_name ) {
+				if (
+					isset( $GLOBALS[ $global_var_name ]['load'] )
+					&&
+					$GLOBALS[ $global_var_name ]['load'] instanceof Closure
+				) {
+					call_user_func( $GLOBALS[ $global_var_name ]['load'] );
+					unset( $GLOBALS[ $global_var_name ] );
+				}
+			};
+
+			// Handle either where the plugin is installed as a regular plugin or is embedded in another plugin or in a theme.
+			if ( ! did_action( 'plugins_loaded' ) ) {
+				add_action( 'plugins_loaded', $bootstrap, 0 );
+			}
+
+			// Handle case where plugin is embedded in a theme.
+			add_action( 'after_setup_theme', $bootstrap, 0 );
+		}
+
+		// Register this copy of the plugin.
+		if (
+			// Register this copy if none has been registered yet.
+			! isset( $GLOBALS[ $global_var_name ]['version'] )
+			||
+			// Or register this copy if the version greater than what is currently registered.
+			version_compare( $version, $GLOBALS[ $global_var_name ]['version'], '>' )
+			||
+			// Otherwise, register this copy if it is actually the one installed in the directory for plugins.
+			rtrim( WP_PLUGIN_DIR, '/' ) === dirname( __DIR__ )
+		) {
+			$GLOBALS[ $global_var_name ]['version'] = $version;
+			$GLOBALS[ $global_var_name ]['load']    = $load;
+		}
+	}
+)(
+	'plsr_pending_plugin_info',
+	'1.2.1',
+	static function () {
+
+		// Define the constant.
+		if ( defined( 'SPECULATION_RULES_VERSION' ) ) {
+			return;
+		}
 
-define( 'SPECULATION_RULES_VERSION', '1.2.1' );
-define( 'SPECULATION_RULES_MAIN_FILE', plugin_basename( __FILE__ ) );
+		define( 'SPECULATION_RULES_VERSION', '1.2.1' );
```

Comment:
It would be great if we could avoid having to modify the version number in more than one place. Could this be defined from the global?

Reply:
Like I removed in bdf96a652d3f6de3a4e26cc52055289327c87977? I thought maybe this was overly fancy and made it more difficult to reason about. Since there's already two instances of the version in the file, a third instance doesn't seem terrible. They should all probably get updated programmatically anyway.

Reply:
Alternatively we could do something like this:

```diff
--- a/plugins/speculation-rules/load.php
+++ b/plugins/speculation-rules/load.php
@@ -20,6 +20,8 @@ if ( ! defined( 'ABSPATH' ) ) {
 	exit;
 }
 
+require_once ABSPATH . 'wp-admin/includes/plugin.php'; // Note: Removed in build step.
+
 (
 	/**
 	 * Register this copy of the plugin among other potential copies.
@@ -62,7 +64,7 @@ if ( ! defined( 'ABSPATH' ) ) {
 	}
 )(
 	'plsr_pending_plugin_info',
-	'1.2.2',
+	get_plugin_data( __FILE__ )['Version'],
 	static function () {
 
 		// Define the constant.
@@ -70,7 +72,7 @@ if ( ! defined( 'ABSPATH' ) ) {
 			return;
 		}
 
-		define( 'SPECULATION_RULES_VERSION', '1.2.1' );
+		define( 'SPECULATION_RULES_VERSION', get_plugin_data( __FILE__ )['Version'] );
 		define( 'SPECULATION_RULES_MAIN_FILE', plugin_basename( __FILE__ ) );
 
 		require_once __DIR__ . '/class-plsr-url-pattern-prefixer.php';
```

And the build step could remove those calls with the actual version. But since this requires _requiring_ a file from `wp-admin/includes` this is dangerous as we could assume the functions in there are present only to have them throw a fatal when called in a production build.

Reply:
It doesn't look like it because the `$global_var_name` is not passed to the `$load` callback. I'll just undo bdf96a6. 

Done: 36fbaf4d